AX: accessibilityAttributeNames does several unnecessary NSArray copies and ancestry traversals

twilco opened the pull request:
#### e6f95871da78d511ad2bdb4a1bb135818288d4d4
<pre>
AX: accessibilityAttributeNames does several unnecessary NSArray copies and ancestry traversals
<a href="https://bugs.webkit.org/show_bug.cgi?id=284212">https://bugs.webkit.org/show_bug.cgi?id=284212</a>
<a href="https://rdar.apple.com/141080358">rdar://141080358</a>

Reviewed by Chris Fleizach.

Prior to this commit, -[WebAccessibilityObjectWrapperMac accessibilityAttributeNames] would unnecessarily call
arrayByAddingObjectsFromArray several times, which makes a copy of the NSArray (at least 2 unnecessary copies were made
for every object, with as many as 4 possible depending on the ARIA attributes and role combination). With this commit,
we are guaranteed to do at most 1, and significantly more often will not need to do any copy at all (e.g. for static
text objects, which are obviously one of the most common types). This saves 230 samples on a slow webpage.

This patch further optimizes -[WebAccessibilityObjectWrapperMac accessibilityAttributeNames] by combining three separate
ancestry traversals into one. This saves ~400 samples on a slow webpage.

Finally, this patch optimizes AXIsolatedObject creation. Prior to this patch, computing AXPropertyName::SortDirection
did an ancestry traversal for every single object in search of a row or column header, the only objects that are valid
to have aria-sort. With this commit, each object only checks its own role and aria-sort presence, saving an ancestry traversal.
Furthermore, this decreases memory usage, as prior to this patch, every descendant of a row or column header with aria-sort
would store that aria-sort value (since every object iterated upwards to find a sort-direction-having ancestor).

* LayoutTests/accessibility/display-contents/role-row-headers-expected.txt:
* LayoutTests/accessibility/image-link-expected.txt:
* LayoutTests/accessibility/image-map2-expected.txt:
* LayoutTests/accessibility/mac/aria-columnrowheaders-expected.txt:
* LayoutTests/accessibility/mac/bounds-for-range-expected.txt:
* LayoutTests/accessibility/mac/document-attributes-expected.txt:
* LayoutTests/accessibility/mac/document-links-expected.txt:
* LayoutTests/accessibility/mac/internal-link-anchors-expected.txt:
* LayoutTests/accessibility/table-attributes-expected.txt:
* LayoutTests/accessibility/table-cell-spans-expected.txt:
* LayoutTests/accessibility/table-cells-expected.txt:
* LayoutTests/accessibility/table-multiple-tbodies-expected.txt:
* LayoutTests/accessibility/table-thead-tfoot-expected.txt:
* LayoutTests/accessibility/table-with-rules-expected.txt:
* LayoutTests/platform/mac/accessibility/lists-expected.txt:
AXBusy is reordered in attribute name lists.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::hasPopup const):
(WebCore::AXCoreObject::selfOrAncestorLinkHasPopup const):
(WebCore::AXCoreObject::sortDirectionIncludingAncestors const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isInsideLiveRegion const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::sortDirection const):
(WebCore::AccessibilityObject::buttonRoleType const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::states const):
(WebCore::AccessibilityObjectAtspi::attributes const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityHasPopup]):
(-[WebAccessibilityObjectWrapper accessibilitySortDirection]):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper _additionalAccessibilityAttributeNames:]):
(-[WebAccessibilityObjectWrapper ALLOW_DEPRECATED_IMPLEMENTATIONS_END]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
(WebCore::InspectorAuditAccessibilityObject::getComputedProperties):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):

Canonical link: <a href="https://commits.webkit.org/287686@main">https://commits.webkit.org/287686@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa458f5eb4769d36164824b785fff71b7f262bea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79884 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58882 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84399 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30865 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81993 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67947 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7174 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62430 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20263 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82952 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52484 "Found 1 new test failure: media/audioSession/getUserMedia.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72751 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42741 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49827 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26900 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29321 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70949 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27383 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85831 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7103 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4976 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70706 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7280 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69948 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17546 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13944 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12873 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7066 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12616 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6917 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10428 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8722 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -1301,12 +1329,8 @@ - (NSArray *)accessibilityAttributeNames
             objectAttributes = tableRowAttrs.get().get();
-    } else if (backingObject->isTreeItem()) {
-        if (backingObject->supportsCheckedState())
-            objectAttributes = [outlineRowAttrs.get() arrayByAddingObject:NSAccessibilityValueAttribute];
```
do we not need this one anymore? couldn't tell if it was replaced by something somewhere else

twilco replied:
We still maintain this logic! It's just been moved to `_additionalAccessibilityAttributeNames`. By moving it to that function, we're grouping all the things that would require a copy to be made via `arrayByAddingObject`. Otherwise, e.g. for tree items that `supportsCheckedState`, we would make a copy here, and then make another copy after calling `_additionalAccessibilityAttributeNames`, which is unnecessary.

(We also used to make another unconditional copy after that to add `NSAccessibilityPrimaryScreenHeightAttribute` for every object).

Ultimately, this guarantees that we only ever make one copy of the attribute NSArray for objects with more uncommon states, e.g. the presence of `aria-checked` on a tree item.

hoffmanjoshua commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -887,30 +882,45 @@ - (NSArray *)_additionalAccessibilityAttributeNames:(const RefPtr<AXCoreObject>&
+    bool addedAriaAtomic = false;
+    bool addedSortDirection = false;
+    for (RefPtr ancestor = backingObject; ancestor; ancestor = ancestor->parentObject()) {
+        if (!addedHasPopup && ancestor->isLink() && ancestor->hasPopup()) {
```
To avoid needing to keep behavior in sync, should we consider changing `selfOrAncestorLinkHasPopup` to return an `AXCoreObject` pointer that we can bool/null check?

twilco replied:
Hmm, it would be great if we could avoid needing to keep the behavior in sync manually (though I think it's a fine cost to pay considering how hot this function is, if we can't figure something out).

You're suggesting something like `selfOrAncestorLinkWithPopup`. How would we use that here? If we just called it outright, we would do an ancestry traversal as part of that function, and then do another to try to find aria-atomic and aria-sort ancestors, and duplicated ancestry traversals is what this PR is trying to avoid since this function can be so frequently called by ATs. But maybe you have a different idea for using this that I'm missing!

hoffmanjoshua replied:
That's a good point—the diff wasn't clear to me that those were all contained within this traversal.

I don't think there is another option then to avoid manually syncing these. What you have now is fine

twilco committed:
AX: accessibilityAttributeNames does several unnecessary NSArray copies and ancestry traversals
https://bugs.webkit.org/show_bug.cgi?id=284212
rdar://141080358

Reviewed by Chris Fleizach.

Prior to this commit, -[WebAccessibilityObjectWrapperMac accessibilityAttributeNames] would unnecessarily call
arrayByAddingObjectsFromArray several times, which makes a copy of the NSArray (at least 2 unnecessary copies were made
for every object, with as many as 4 possible depending on the ARIA attributes and role combination). With this commit,
we are guaranteed to do at most 1, and significantly more often will not need to do any copy at all (e.g. for static
text objects, which are obviously one of the most common types). This saves 230 samples on a slow webpage.

This patch further optimizes -[WebAccessibilityObjectWrapperMac accessibilityAttributeNames] by combining three separate
ancestry traversals into one. This saves ~400 samples on a slow webpage.

Finally, this patch optimizes AXIsolatedObject creation. Prior to this patch, computing AXPropertyName::SortDirection
did an ancestry traversal for every single object in search of a row or column header, the only objects that are valid
to have aria-sort. With this commit, each object only checks its own role and aria-sort presence, saving an ancestry traversal.
Furthermore, this decreases memory usage, as prior to this patch, every descendant of a row or column header with aria-sort
would store that aria-sort value (since every object iterated upwards to find a sort-direction-having ancestor).

* LayoutTests/accessibility/display-contents/role-row-headers-expected.txt:
* LayoutTests/accessibility/image-link-expected.txt:
* LayoutTests/accessibility/image-map2-expected.txt:
* LayoutTests/accessibility/mac/aria-columnrowheaders-expected.txt:
* LayoutTests/accessibility/mac/bounds-for-range-expected.txt:
* LayoutTests/accessibility/mac/document-attributes-expected.txt:
* LayoutTests/accessibility/mac/document-links-expected.txt:
* LayoutTests/accessibility/mac/internal-link-anchors-expected.txt:
* LayoutTests/accessibility/table-attributes-expected.txt:
* LayoutTests/accessibility/table-cell-spans-expected.txt:
* LayoutTests/accessibility/table-cells-expected.txt:
* LayoutTests/accessibility/table-multiple-tbodies-expected.txt:
* LayoutTests/accessibility/table-thead-tfoot-expected.txt:
* LayoutTests/accessibility/table-with-rules-expected.txt:
* LayoutTests/platform/mac/accessibility/lists-expected.txt:
AXBusy is reordered in attribute name lists.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::hasPopup const):
(WebCore::AXCoreObject::selfOrAncestorLinkHasPopup const):
(WebCore::AXCoreObject::sortDirectionIncludingAncestors const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isInsideLiveRegion const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::sortDirection const):
(WebCore::AccessibilityObject::buttonRoleType const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::states const):
(WebCore::AccessibilityObjectAtspi::attributes const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityHasPopup]):
(-[WebAccessibilityObjectWrapper accessibilitySortDirection]):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper _additionalAccessibilityAttributeNames:]):
(-[WebAccessibilityObjectWrapper ALLOW_DEPRECATED_IMPLEMENTATIONS_END]):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
(WebCore::InspectorAuditAccessibilityObject::getComputedProperties):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):

Canonical link: https://commits.webkit.org/287686@main
